generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -672,9 +672,52 @@ public Expr apply(Expr expr, BasicType type, boolean constContext, int depth,
       // (false ? whatever : expr)
       // or
       // (true  ? expr : whatever)
-      assert BasicType.allScalarTypes().contains(type);
-      // Only generate ternary expressions for scalars; the vector case causes massive blow-up
-      Expr exprWithIdentityApplied = applyIdentityFunction(expr, type, constContext, depth, fuzzer);
+
+      // If we generate identities for all entries in a non-scalar type, a huge amount of
+      // identities are applied. To solve that, we only fuzz one entry at random in a vector,
+      // and one column at random in a matrix.
+      Expr exprWithIdentityApplied;
+      if (!BasicType.allScalarTypes().contains(type)) {
+        // v + vec2(0.0)[0]; is not valid GLSL
+        if (!(expr instanceof BinaryExpr)) {
+          int numColumns;
+          int indexToApplyIdentities;
+          if (!BasicType.allMatrixTypes().contains(type)) {
+            // vector case
+            // v -> (true ? vecX(..., identity(v[Y]), ...) : _)
+            // v -> (false ? _ : vecX(..., identity(v[Y]), ...))
+            // where v is a vector of size X, y is the random entry in v we want to apply
+            // identities to, and ... is the other entries in v that we don't change.
+
+            numColumns = type.getNumElements();
+            indexToApplyIdentities = generator.nextInt(numColumns);
+
+            List<Expr> vectorEntryList = new ArrayList<Expr>();
+            for (int i = 0; i < numColumns; i++) {
+              vectorEntryList.add(new ArrayIndexExpr(expr.clone(),
+                  new IntConstantExpr(String.valueOf(i))));
+            }
+            vectorEntryList.set(indexToApplyIdentities,
```

Comment:
@afd This can lead to a blow up if the expression (of vector type) is already big, as it is replicated e.g. 4 times.
TODO: follow-up on best approach to cope with this.